### ISSUE TITLE
Poloniex getting wallet says Nonce must be greater than 1581924041151. You provided 899805527

### ISSUE THREAD

**Issue by andrewlin0410** [knowm/XChange#3406]
The first number seems to be always the same and the second seems to go up by about 1 every second. I am basically getting wallet in a loop that goes through every few seconds.

`         

while (true) {            // every few seconds 
  poloniexWallet = poloniexExchange.getAccountService().getAccountInfo().getWallet(); // 
}
`

poloniexExcahnge is of class Exchange and initialized by this following: 
`Exchange createPoloniexExchange() {
        ExchangeSpecification exSpec = new PoloniexExchange().getDefaultExchangeSpecification();
        exSpec.setApiKey("api"); // original poloniex
        exSpec.setSecretKey("secret");
        return ExchangeFactory.INSTANCE.createExchange(exSpec);
    }`

**Comment by andrewlin0410** [comment]
I think the first number is a timestamp in the recent hour but the I am not sure what the second number is. If it were a timestamp, it would be in 1998 it seems

**Comment by andrewlin0410** [comment]
So does anyone have the same issue? How do you fix it? Thanks

**Comment by mdvx** [comment]
1581924041151:
GMT: Monday, February 17, 2020 7:20:41.151 AM
Your time zone: Monday, February 17, 2020 2:20:41.151 AM GMT-05:00
Relative: An hour ago

PoloniexExchange:
private SynchronizedValueFactory<Long> nonceFactory = new TimestampIncrementingNonceFactory();

  private static final long START_MILLIS =
      1356998400000L; // Jan 1st, 2013 in milliseconds from epoch

    lastNonce = Math.max(lastNonce + 1, (int) ((System.currentTimeMillis() - START_MILLIS) / 250L));

**Comment by mdvx** [comment]
ok, so one new Nonce every 250ms, and poloniex is telling you nonce is too small

to fix: either / or
a) create a new API key, and dont query it this a none greater than TimestampIncrementingNonceFactory can create
b) override PoloniexExchange createNonceFactory() and 
   return new AtomicLongCurrentTimeIncrementalNonceFactory()

**Comment by andrewlin0410** [comment]
Hi thanks for the reply but I am still not sure how I can apply the fix to my code. Can you please point the way?

**Comment by mdvx** [comment]
something like this should do it:

`class MyPoloniexExchange extends PoloniexExchange {  // or PoloniexStreamingExchange
     private final SynchronizedValueFactory<Long> nonceFactory = new AtomicLongCurrentTimeIncrementalNonceFactory();
     @Override
    public SynchronizedValueFactory<Long> getNonceFactory() {
        return nonceFactory;
    }
}`

`Exchange xchange = ExchangeFactory.INSTANCE.createExchange(MyPoloniexExchange .class, exSpec);`

**Comment by andrewlin0410** [comment]
I don't see an implementation of createExchange with (Class<T>, ExchangExchangeSpecification) though :/

**Comment by andrewlin0410** [comment]
Would this work? 
'ExchangeSpecification exSpec = new MyPoloniexExchange().getDefaultExchangeSpecification();
        exSpec.setApiKey("api"); // poloniex f
        exSpec.setSecretKey("secret");
        return ExchangeFactory.INSTANCE.createExchange(exSpec);'

**Comment by mdvx** [comment]
No, not unless you override getDefaultExchangeSpecification() too
try:
 `return ExchangeFactory.INSTANCE.createExchange(MyPoloniexExchange.class, apiKey, ApiSecret);`

**Comment by andrewlin0410** [comment]
This throws this exception 
org.knowm.xchange.exceptions.ExchangeException: Problem creating Exchange (illegal access)

**Comment by mdvx** [comment]
you will need a public default constructor for MyPoloniexExchange

**Comment by andrewlin0410** [comment]
It does not work. Also I saw that the default constructor is automatically provided if one is not provided. 
Anyways I tried to add the following 
`public MyPoloniexExchange () {
     }`
or
`public MyPoloniexExchange () {
          super();
     }`
neither worked

**Comment by andrewlin0410** [comment]
Forgive me if I made a mistake I don't really know constructor well enough

**Comment by mdvx** [comment]
I use:

`public MyPoloniexExchange () { super(MyPoloniexExchange.class); }`

**Comment by andrewlin0410** [comment]
Unfortunately, this gives me expected 0 arguments but found 1

**Comment by mdvx** [comment]
You need to find a friend who can help you with the Java side, you have a lot of information in this thread, to help them.

**Comment by RomanHiden** [comment]
I have dealt with poloniex myself with exactly the same issue. For me it was just creating another api key on poloniex side. so two computers would use different api keys. then nounce issue dissapeard
